python/ray/train/ml_trainer.py

```diff
@@ -0,0 +1,117 @@
+import abc
+import copy
+import logging
+from typing import Dict, Union, Callable, Optional, TYPE_CHECKING
+
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.result import Result
+from ray import tune
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.tune.utils import deep_update
+
+if TYPE_CHECKING:
+    from ray.data import Dataset
+
+GenDataset = Union["Dataset", Callable[[], "Dataset"]]
+
+
+logger = logging.getLogger(__name__)
+
+
+class Trainer(ConvertibleToTrainable, abc.ABC):
```

Comment:
I think we should spend a little more time thinking about how we expect users to interface with this class. I'm not talking about passing in all the configuration dicts, but more how a Trainer would be implemented in practice. I'm assuming we have a lot of default implementations in place, but I'd argue that we should still make it obvious how to extend a `Trainer` for a new, fancy framework.

In principle, all users have to do is to implement `as_trainable`, which seems easy enough at first, but introduced a million obstacles down the line. Users either need to know a) how to define a `tune.Trainable`, i.e they have to know that they need to implement a `step`, `save_checkpoint` and `from_checkpoint` method, or b) how to specify a function which can be wrapped as `FunctionRunner` with the right tooling. But this is all quite a lot to ask.

Can we e.g. instead have a method called `training_step`, which returns a function and we simply use `wrap_function` internally to make this a `FunctionRunner`? Would we lose a lot by restricting ourselves to functions (instead of full-blown Trainables)? I think a `Trainer` that has to simply define a `training_step` is a _lot_ more intuitive then what we're proposing currently.

Reply:
I couldn't agree more for this. I even don't think the public doc is finalized in OSS.

Reply:
In the original prototype, we implemented a `FunctionTrainer` where the training function could easily be defined:

https://github.com/krfricke/ray/blob/tune/api-prototype-rebase/python/ray/tune/examples/tuner_api.py#L37-L72

We could either keep this approach or extend the current Trainer interface to use this per default. The user then never has to define `as_trainable()`, but only the function trainable, as described by @maxpumperla.

Additionally this makes it easy for us to expose parameters like the scaling config to the user training functions.